### PR TITLE
change stride to 4 instead of 1 if int8 is not supported

### PR DESCRIPTION
**Pull request by Rekt3421**
please see issue #1396 for more details

### CODE REVIEW

**Comment by Rekt3421** [comment]
Is a unit test needed for this? Seems a little trivial @rjodinchr ?

**Review by rjodinchr — CHANGES_REQUESTED** [review]
This is not the only place where we create `spv::DecorationArrayStride`.
I would like to have every place fixed.

Also, why `2`? when int8 is not supported, it is converted to int32 not int16.

We would definitely need tests here. At least the one from  the issue.

**Comment by rjodinchr** [comment]
This one feels like a good one as well:
```
// RUN: clspv -int8=0 %target %s -o %t.spv
// RUN: spirv-dis -o %t2.spvasm %t.spv
// RUN: FileCheck %s < %t2.spvasm
// RUN: spirv-val --target-env vulkan1.0 %t.spv

constant uchar b[4] = {[0]=42, [1]=13, [2]=0, [3]=5};

void kernel __attribute__((reqd_work_group_size(4, 1, 1))) foo(global uchar* a)
{
  *a = b[get_local_id(0)];
}

// CHECK-DAG:  [[uint:%[^ ]+]] = OpTypeInt 32 0
// CHECK-DAG:  OpTypePointer StorageBuffer [[struct:%[^ ]+]]
// CHECK-DAG:  [[struct]] = OpTypeStruct [[runtimearr:%[^ ]+]]
// CHECK-DAG:  OpDecorate [[runtimearr]] ArrayStride 4
```
You will note that even if you change `ArrayStride 4` to `ArrayStride 2` in the test, it is still not passing with 2. It would be failing with the following message:
```
Structure id 15 decorated as Block for variable in StorageBuffer storage class must follow standard storage buffer layout rules: member 0 contains an array with stride 2 not satisfying alignment to 4
```

**Comment by Rekt3421** [comment]
Will add the test from the issue as well.

File: lib/SPIRVProducerPass.cpp [hunk 1]
```diff
@@ -2084,8 +2084,12 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVType(Type *Ty, bool needs_layout) {
         // Ops[2] = Stride Number(Literal Number)
         Ops.clear();
 
-        Ops << RID << spv::DecorationArrayStride
-            << static_cast<uint32_t>(GetTypeAllocSize(EleTy, DL));
+        // if stride is 1 automatically set it to 2 to avoid stride issues.
+        auto CurrStride = static_cast<uint32_t>(GetTypeAllocSize(EleTy, DL));
+        if (CurrStride == 1 && !Int8Support()) {
+          CurrStride = 4;
+        }
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:2091]: Please avoid code duplication. Make a function for it.

**Review by rjodinchr — CHANGES_REQUESTED** [review]

File: lib/SPIRVProducerPass.cpp [hunk 2]
```diff
@@ -2084,8 +2091,10 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVType(Type *Ty, bool needs_layout) {
         // Ops[2] = Stride Number(Literal Number)
         Ops.clear();
 
+        // if stride is 1 automatically set it to 2 to avoid stride issues.
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:2094]: The comment is not correct anymore, please update it.

Comment (Rekt3421) [reply lib/SPIRVProducerPass.cpp]: thank you!

**Comment by rjodinchr** [comment]
The `macos` build are failing because this branch has not been rebase on ToT.

**Pull request merged**